pika/adapters/select_connection.py

```diff
@@ -188,23 +135,54 @@ class SelectPoller(object):
     to override the update_handler and start methods for additional types.
 
     """
-    TIMEOUT = 0.01
+    # Drop out of the poll loop every POLL_TIMEOUT secs as a worst case, this
+    # is only a backstop value. We will run timeouts when they are scheduled.
+    POLL_TIMEOUT = 5
+    # if the poller uses MS specify 1000
+    POLL_TIMEOUT_MULT = 1
 
-    def __init__(self, fileno, handler, events, state_manager):
+    def __init__(self):
         """Create an instance of the SelectPoller
 
-        :param int fileno: The file descriptor to check events for
-        :param method handler: What is called when an event happens
-        :param int events: The events to look for
-        :param method state_manager: The method to manage state
-
         """
-        self.fileno = fileno
-        self.events = events
-        self.open = True
-        self._handler = handler
-        self._timeouts = []
-        self._manage_event_state = state_manager
+        self._fd_handlers = dict()
+        self._fd_events = {READ: set(), WRITE: set(), ERROR: set()}
+        self._stopping = False
+        self._timeouts = {}
+        self._next_timeout = None
+        self._processing_fd_event_map = {}
+        self._r_interrupt, self._w_interrupt = self.get_interrupt_pair()
+        self.add_handler(self._r_interrupt.fileno(), self.read_interrupt, READ)
+
+    def get_interrupt_pair(self):
+        """ Use a socketpair to be able to interrupt the ioloop if called
+            from another thread. Socketpair() is not supported on some OS (Win)
+            so use a pair of simple UDP sockets instead. The sockets will be
+            closed and garbage collected by python when the ioloop itself is.
+        """
+        try:
+            return socket.socketpair()
+
+        except:
+            LOGGER.debug("Using custom socketpair for interrupt")
+            read_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            read_sock.setblocking(0)
+            read_sock.bind(('localhost', 0))
+            write_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            write_sock.setblocking(0)
+            write_sock.connect(read_sock.getsockname())
+            return read_sock, write_sock
+
+    def read_interrupt(self, interrupt_sock, events, write_only):
+        """ Read the interrupt byte(s). We ignore the event mask and write_only
+        flag as we can ony get here if there's data to be read on our fd.
+
+        :param int interrupt_sock: The file descriptor to read from
+        :param int events: (unused) The events generated for this fd
+        :param bool write_only: (unused) True if poll was called to trigger a
+            write 
+        """
+        os.read(interrupt_sock, 512)
```

Comment:
@wjps, this may be a bug. Since `get_interrupt_pair()` configured the socket pair as non-blocking, OSError exception with errno EAGAIN or EWOULDBLOCK may be a possibility